chat-sdk-core/src/main/java/co/chatsdk/core/base/BaseLocationAdapter.java

```
@@ -0,0 +1,104 @@
+    }
+
+    public Observable<Location> requestLocationUpdates(long interval, int distance) {
+        return requestLocationUpdates(interval)
+                .distinctUntilChanged((l1, l2) -> l1.distanceTo(l2) < distance);
+    }
+
+    public Observable<Location> requestLocationUpdates(long interval) {
+        LocationRequest locationRequest = new LocationRequest();
+        locationRequest.setInterval(interval * 1000);
+        locationRequest.setFastestInterval(interval * 1000);
+        locationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);
+        if (ContextCompat.checkSelfPermission(context(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(context(), Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+            return Observable.error(new Error("location permissions not granted"));
+        }
+        return Observable.create(observable -> {
+            LocationCallback locationCallback = new LocationCallback() {
+                @Override
+                public void onLocationResult(LocationResult locationResult) {
+                    Location location = getMostAccurateLocation(locationResult.getLocations());
+                    if (location != null) {
+                        observable.onNext(location);
+                    }
+                }
+            };
+            locationClient.requestLocationUpdates(locationRequest, locationCallback, Looper.myLooper());
+        });
+    }
+
+    public Single<Location> getLastLocation() {
+        if (ContextCompat.checkSelfPermission(context(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(context(), Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+            return Single.error(new Error("location permissions not granted"));
+        }
```

Maybe move this request permission to a new function so the same code isn't duplicated.

---

Again, all the code needs to be wrapped in the `Single.create` block.

---

I inlined this permissions check code to disable the the missing permissions error from Android Studio

---

If you want to silence the error,  you can do that by [suppressing the warning](https://stackoverflow.com/questions/17603677/supported-suppresswarnings-values-in-android-studio):